[Step4] Team Sugan

SunnnySong opened the pull request:
> 안녕하세요! 코리(@corykim0829) Team Sugan: Beyond the Livers 의 Sunny와 Logan 입니다!!⛵️ 😆 이번 마지막 PR로 4주간의 항해를 마무리 하려고 합니다..! 요청하신대로 아직 미완성이지만 진행된 만큼이라도 PR 을 보내게 됐습니다. 아직 미완성이라도 좋은 리뷰 남겨주셨으면 좋겠습니다..!


### 구현 
- Timer 구현
처음 구현은 Timer을 viewController에서 구현했었어요. 하지만 점점 viewController가 비대해지고 있다는 생각이 들어 TimerModel이라는 파일로 분리했어요. 

-  UI 구현
Step4에 대한 UI의 기본 구현은 완료했지만 기능을 연결하지 못했어요. 하지만 custom으로 UILabel과 UIStackView등을 구현해보았습니다. viewController에서 모든 UI object에 대한 속성을 정의하기엔 코드가 너무 길어지기 때문에 이런 방식을 시도해 보았습니다. 

- 아직 미완성이라 구현하진 않았지만 completionHandler를 Console버전에서는 출력하는데 사용했다면 이번에는 "대기중"과 "업무중"의 StackView에 커스텀된 라벨들을 생성 및 제거하는 역할을 맡기려고 합니다.

---
### Q: 궁금한점 🤔
- 이번 프로젝트에서 처음으로 모든 UI를 코드로만 구현해봤습니다. 구현하는 과정에서 생각보다 코드로 UI로 구현하는것이 많이 번거롭다는것을 새롭게 깨달았어요. 여기서 궁금한 건 각 UI오브젝트들의 Layout을 잡는 과정에서 우리가 코드들의 역할을 정리하듯이 각 Layout을정리하는 메서드를 나누는 기준이 생각보다 애매한 감이 없지 않아 있더라구요. 현업에서는 어떤 기준으로 나누는지 궁금합니다. (예를들어 하나의 UI오브젝트 기준으로 나누는지, 아니면 Layout의 기능을 기준으로 나누는지 정도가 되겠네요)

- 이번 프로젝트에서 UI object에 대한 속성을 선언하기 보단, custom 으로 만들어서 사용했었어요. 또한 Color라는 asset 을 만들어 처음 사용해보았어요. 이러한 방식이 좋은 구현 방법인지, 현업에서 자주 사용하는 방식인지 궁금합니다.😃

sloDevz committed:
BankManager TargetMembership 헤제

sloDevz committed:
feat: 고객추가,초기화버튼 구현, 타이머 라벨, 업무현황 라벨 UI 구현

SunnnySong committed:
refactor: makeTeller 메서드 생성

sloDevz committed:
feat: CustomerInfoView 이름으로 커스텀 뷰 파일 생성

sloDevz committed:
Merge branch 'loganDev4'

SunnnySong committed:
feat: Palette 생성 및 UIColor extension 생성

sloDevz committed:
feat: 고객 현황판 스택 생성

sloDevz committed:
merge from main

SunnnySong committed:
refactor: autoLayout 수정

SunnnySong committed:
feat: BankTextCollection 생성

SunnnySong committed:
feat: BankTimer 생성

sloDevz committed:
feat: 각 요소들 StackView로 묶음

sloDevz committed:
merge loganDev4 to main

sloDevz committed:
chore: time 메서드 private

SunnnySong committed:
feat: CustomStackView 생성 및 적용

SunnnySong committed:
feat: CustomStackView 생성 및 적용

SunnnySong committed:
refactor: CustomStackView init 수정

sloDevz committed:
feat: waiting, inprogress stackView에 CustomerInfoView 생성구현
- waitingStack에 6개 이상 안들어가는 점 고쳐야해요
- CustomerInfoView 안에있는 라벨들의 폭이 글자 폭만큼 되도록 구현해야해요

sloDevz committed:
머지 취소

SunnnySong committed:
Merge branch 'loganDev4'

sloDevz committed:
Merge branch 'main' of https://github.com/SunnnySong/ios-bank-manager

SunnnySong committed:
refactor: waitingStackView, inprogressStackView 수정

sloDevz committed:
Merge branch 'main' of https://github.com/SunnnySong/ios-bank-manager

sloDevz committed:
chore: waiting, inprogress stack topanchor 15로 변경

SunnnySong committed:
refactor: BasicStackView 이름 변경, waitingStackView/inprogressStackView에 활용

SunnnySong committed:
Merge branch 'main' of https://github.com/SunnnySong/ios-bank-manager

SunnnySong committed:
feat: Extension+UIStackView 추가 및 적용

SunnnySong committed:
refactor: CustomerInfoView의 infoStackView 수정

SunnnySong committed:
refactor: BasicStackView의 translatesAutore~ 설정 포함하게 수정

SunnnySong committed:
chore: 주석 및 테스트 코드 삭제

sloDevz committed:
chore: BankManger 위치 이동

sloDevz committed:
chore: Console버전 레퍼런스 참조

SunnnySong committed:
chore: target membership 수정

SunnnySong committed:
feat: generateWaiting 함수 수정

SunnnySong committed:
feat: BasicLabel 생성

SunnnySong committed:
feat: TimerModel 생성, ViewController에서 Timer 분리

sloDevz committed:
feat: generateWaiting 메서드 수정후 로직에 적용
- generateWaiting메서드에게 Queue에 넣는일, Customer배열을 만드는 역할 부여
- 만들어진 Customer배열을 리턴하게 하여 ViewController에서 받아서 UI에 업데이트

sloDevz committed:
feat: merge main with loganDev4

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/DesignSystem/CustomView/BasicLabel.swift:
```diff
@@ -0,0 +1,32 @@
+        super.init(frame: frame)
+    }
+    
+    convenience init(systemFontSize: CGFloat) {
+        self.init(frame: .zero)
+        
+        self.font = UIFont.systemFont(ofSize: systemFontSize)
+    }
+    
+    convenience init(systemFontSize: CGFloat, textAlignnment: NSTextAlignment) {
```
UILabel의 textAlignment는 default값이 있기 때문에 그 값을 생성자 파라미터로 처리하면 두개의 생성자를 하나로 묶을 수 있을 것 같네요!

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/DesignSystem/CustomView/BasicStackView.swift:
```diff
@@ -0,0 +1,39 @@
+        setupConfiguration()
+    }
+    
+    convenience init(axis: NSLayoutConstraint.Axis) {
+        self.init(frame: .zero)
+        self.axis = axis
+    }
+    
+    convenience init(axis: NSLayoutConstraint.Axis, alignment: UIStackView.Alignment) {
```
여기도 BasicLabel과 마찬가지로 생성자 하나로 처리할 수 있을 것 같아요!

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/ViewController.swift:
```diff
@@ -7,12 +7,173 @@
 }
 
+
+// MARK: - Timer
```
TimerDelegate의 구현부분이니 주석에 TimerDelegate라고 하면 더 명확하고 좋을 것 같네요!

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/BankModel/TimerModel.swift:
```diff
@@ -0,0 +1,47 @@
+
+final class TimerModel {
+    
+    weak var timerDelegate: TimerDelegate?
```
이미 클래스명이 timer라는 단어가 들어가서 timerDelegate라고 하면 중복이 되어서 delegate라고만 처리해도 좋을 것 같습니다!
retain cycle 방지를 위한 weak var로 처리한 부분 좋습니다 👍

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/BankModel/TimerModel.swift:
```diff
@@ -0,0 +1,47 @@
+    
+    weak var timerDelegate: TimerDelegate?
+    
+    weak var timer: Timer?
```
여기 timer는 weak var로 처리한 이유가 있나요?

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/DesignSystem/CustomView/CustomerInfoView.swift:
```diff
@@ -0,0 +1,52 @@
+        
+        infoStackView.addArrangedSubviews([ticketNumber, hyphenLabel, taskTypeLabel])
+        
+        infoStackView.translatesAutoresizingMaskIntoConstraints = false
```
BasicStackView 에서도 translatesAutoresizingMaskIntoConstraints를 false 처리하고 있어 중복되네요. BasicStackView인 만큼 너무 세세한 부분까지 BasicStackView에서 많이 처리하면 추후에 헷갈릴 수 있으니 이러한 처리는 외부에서 처리하는게 더 깔끔하고 실수를 줄일 수 있으니 참고하세요

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/Extension/Extension+StackView.swift:
```diff
@@ -0,0 +1,17 @@
+
+extension UIStackView {
+    
+    func addArrangedSubviews(_ views: [UIView]) {
```
extension 구현 깔끔하네요! 👍

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/ViewController.swift:
```diff
@@ -7,12 +7,173 @@
 
+    //MARK: - Properties: Button
+    
+    private lazy var addCustomersButton: UIButton = {
```
클로저를 사용한 UI 요소 구현에서 
일부 UI 요소들은 lazy var, 아래 일부 라벨은 let으로 구현한 이유가 있나요?

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/ViewController.swift:
```diff
@@ -7,12 +7,173 @@
+    
+    // MARK: - Properties: Timer
+    
+    private let timer = TimerModel()
```
타이머 분리해서 구현한 부분 아주 깔끔하고 좋습니다! 👍

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/ViewController.swift:
```diff
@@ -7,12 +7,173 @@
 
 class ViewController: UIViewController {
+    
+    private let queue = WaitingQueue<Customer>()
```
그냥 큐라고 하기보다는 waitingQueue라고 하는게 의미를 더 명확하게 표현할 수 있을 것 같네요

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/ViewController.swift:
```diff
@@ -7,12 +7,173 @@
+        view.addSubview(waitingStackView)
+        view.addSubview(inprogressStackView)
+        
+        setButtonStack()
+        setTimerLabelStackView()
+        setProgressStackView()
+        setInterfaceStackView()
+        setCustomerWaitingStack()
+        setCustomerInprogressStack()
```
setButtonStackLayout() 와 같이 함수명을 지으면 어떤 일을 하는지 명확하게 파악할 수 있을 것 같아요.

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/DesignSystem/CustomView/CustomerInfoView.swift:
```diff
@@ -0,0 +1,52 @@
+
+import UIKit
+
+final class CustomerInfoView: UIView {
```
CustomerInfoView를 따로 구현하신 부분 좋습니다 👍

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/ViewController.swift:
```diff
@@ -7,12 +7,173 @@
+    
+    private let timerLabelStackView = BasicStackView(axis: .horizontal)
+
+    //MARK: - Properties: StateLabels & Stack
```
프로퍼티 주석은 주석의 내용이 프로퍼티명에 있으므로 프로퍼티 시작점을 알려주는게 아니라면 불필요해보이네요

corykim0829 commented on BankManagerUIApp/BankManagerUIApp/ViewController.swift:
```diff
@@ -7,12 +7,173 @@
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        setUI()
+        setupDelegate()
```
set, setup 도 통일하면 좋을 것 같고

setupDelegate라고 하면 추상적이니 어떤 Delegate를 처리하는지 명시하면 좋을 것 같습니다!

corykim0829 reviewed (approved):
두분 정말 수고 많으셨습니다!! 👍
UI 구현하는 부분이 시간도 많이걸려서 고생을 많이하신 것 같습니다. 

파일 분리도 잘하셨고, UI 구현도 코드로 깔끔하게 잘 구현하셨습니다! 
메소드 분리도 잘하셔서 따봉을 엄청 드리고 싶네요!! 👍👍👍👍👍👍👍

---

Q: 궁금한점 🤔

> 이번 프로젝트에서 처음으로 모든 UI를 코드로만 구현해봤습니다. 구현하는 과정에서 생각보다 코드로 UI로 구현하는것이 많이 번거롭다는것을 새롭게 깨달았어요. 여기서 궁금한 건 각 UI오브젝트들의 Layout을 잡는 과정에서 우리가 코드들의 역할을 정리하듯이 각 Layout을정리하는 메서드를 나누는 기준이 생각보다 애매한 감이 없지 않아 있더라구요. 현업에서는 어떤 기준으로 나누는지 궁금합니다. (예를들어 하나의 UI오브젝트 기준으로 나누는지, 아니면 Layout의 기능을 기준으로 나누는지 정도가 되겠네요)

➡️ 저의 방법이 정답은 아니겠지만 저는 Layout을 처리하는 부분을 한번에 처리하고, 너무 복잡하다면 오브젝트별로 메소드로 나누어 처리했습니다. 
지금은 UI 요소가 많지 않지만, 많아지면 UI 오브젝트들 끼리 Anchor들을 연결하는게 복잡해지기 때문에 복잡하지 않게 처리하는게 좋은 것 같아요.

> 이번 프로젝트에서 UI object에 대한 속성을 선언하기 보단, custom 으로 만들어서 사용했었어요. 또한 Color라는 asset 을 만들어 처음 사용해보았어요. 이러한 방식이 좋은 구현 방법인지, 현업에서 자주 사용하는 방식인지 궁금합니다.😃

➡️ 저도 enum으로 Color Assets을 구현해서 쓰고 있습니다. 자주쓰는 컬러들은 상수화 처리해두면 편리하죠! 잘 구현하셨습니다 👍

sloDevz committed:
chore: TimerDelegate 이름 수정

sloDevz committed:
chore: 분리된 Timer 적용

sloDevz committed:
feat: addLabel 메서드 추가

sloDevz committed:
feat: UIStackView에서 View제거하는 메서드 추가
- 제거할 View가 있는 UIStackView를 입력하고 제거 할 View의 ticketNumber를 입력하면 돼요

sloDevz committed:
chore: 메서드명 오타 수정

sloDevz committed:
feat: 대기열 기능 구현
- Tellers를 전역프로퍼티로 사용하기 위해서 BankManager를 수정했어요
- 대기열에 View들의 추가,삭제를 mian스레드에서 변경하도록 했어요

sloDevz committed:
feat: dealCustomer 메서드 internal로 변경

sloDevz committed:
feat: tellers 전역 프로퍼티로 변
